espidf ledc fails to compile for ESP32C3 board

### The problem

Trying to use LEDC to control a H-Bridge for PWM fan control:

```
output:
  - platform: ledc
    pin: 4
    id: fan1
    frequency: "100Hz"
```

It fails saying that LEDC_HIGH_SPEED_MODE isn't declared.  I do know that the ESP32C3 only has LEDC_LOW_SPEED_MODE, so maybe that's the problem? 

```
src/esphome/components/ledc/ledc_output.cpp: In member function 'virtual void esphome::ledc::LEDCOutput::write_state(float)':
src/esphome/components/ledc/ledc_output.cpp:51:36: error: 'LEDC_HIGH_SPEED_MODE' was not declared in this scope
   auto speed_mode = channel_ < 8 ? LEDC_HIGH_SPEED_MODE : LEDC_LOW_SPEED_MODE;
                                    ^~~~~~~~~~~~~~~~~~~~
src/esphome/components/ledc/ledc_output.cpp:51:36: note: suggested alternative: 'LEDC_LOW_SPEED_MODE'
   auto speed_mode = channel_ < 8 ? LEDC_HIGH_SPEED_MODE : LEDC_LOW_SPEED_MODE;
                                    ^~~~~~~~~~~~~~~~~~~~
                                    LEDC_LOW_SPEED_MODE
src/esphome/components/ledc/ledc_output.cpp: In member function 'virtual void esphome::ledc::LEDCOutput::setup()':
src/esphome/components/ledc/ledc_output.cpp:66:36: error: 'LEDC_HIGH_SPEED_MODE' was not declared in this scope
   auto speed_mode = channel_ < 8 ? LEDC_HIGH_SPEED_MODE : LEDC_LOW_SPEED_MODE;
                                    ^~~~~~~~~~~~~~~~~~~~
src/esphome/components/ledc/ledc_output.cpp:66:36: note: suggested alternative: 'LEDC_LOW_SPEED_MODE'
   auto speed_mode = channel_ < 8 ? LEDC_HIGH_SPEED_MODE : LEDC_LOW_SPEED_MODE;
                                    ^~~~~~~~~~~~~~~~~~~~
                                    LEDC_LOW_SPEED_MODE
src/esphome/components/ledc/ledc_output.cpp: In member function 'virtual void esphome::ledc::LEDCOutput::update_frequency(float)':
src/esphome/components/ledc/ledc_output.cpp:117:36: error: 'LEDC_HIGH_SPEED_MODE' was not declared in this scope
   auto speed_mode = channel_ < 8 ? LEDC_HIGH_SPEED_MODE : LEDC_LOW_SPEED_MODE;
                                    ^~~~~~~~~~~~~~~~~~~~
src/esphome/components/ledc/ledc_output.cpp:117:36: note: suggested alternative: 'LEDC_LOW_SPEED_MODE'
   auto speed_mode = channel_ < 8 ? LEDC_HIGH_SPEED_MODE : LEDC_LOW_SPEED_MODE;
                                    ^~~~~~~~~~~~~~~~~~~~
                                    LEDC_LOW_SPEED_MODE
*** [.pioenvs/datacloset-temp/src/esphome/components/ledc/ledc_output.o] Error 1
```

### Which version of ESPHome has the issue?

2021.12.0-dev20211122

### What type of installation are you using?

Docker

### Which version of Home Assistant has the issue?

_No response_

### What platform are you using?

ESP32-IDF

### Board

ESP32-C3

### Component causing the issue

output/ledc

### Example YAML snippet

_No response_

### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

--- a/esphome/components/ledc/ledc_output.cpp
+++ b/esphome/components/ledc/ledc_output.cpp
@@ -15,6 +15,18 @@ namespace ledc {
 
 static const char *const TAG = "ledc.output";
 
+#ifdef USE_ESP_IDF
+#if SOC_LEDC_SUPPORT_HS_MODE
+// Only ESP32 has LEDC_HIGH_SPEED_MODE
+inline ledc_mode_t get_speed_mode(uint8_t channel) { return channel < 8 ? LEDC_HIGH_SPEED_MODE : LEDC_LOW_SPEED_MODE; }
+#else
+// S2, C3, S3 only support LEDC_LOW_SPEED_MODE
+// See
+// https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-reference/peripherals/ledc.html#functionality-overview
+inline ledc_mode_t get_speed_mode(uint8_t) { return LEDC_LOW_SPEED_MODE; }
+#endif
+#endif
+
 float ledc_max_frequency_for_bit_depth(uint8_t bit_depth) { return 80e6f / float(1 << bit_depth); }
 float ledc_min_frequency_for_bit_depth(uint8_t bit_depth) {
   const float max_div_num = ((1 << 20) - 1) / 256.0f;
@@ -48,7 +60,7 @@ void LEDCOutput::write_state(float state) {
   ledcWrite(this->channel_, duty);
 #endif
 #ifdef USE_ESP_IDF
-  auto speed_mode = channel_ < 8 ? LEDC_HIGH_SPEED_MODE : LEDC_LOW_SPEED_MODE;
+  auto speed_mode = get_speed_mode(channel_);
   auto chan_num = static_cast<ledc_channel_t>(channel_ % 8);
   ledc_set_duty(speed_mode, chan_num, duty);
   ledc_update_duty(speed_mode, chan_num);
@@ -63,7 +75,7 @@ void LEDCOutput::setup() {
   ledcAttachPin(this->pin_->get_pin(), this->channel_);
 #endif
 #ifdef USE_ESP_IDF
-  auto speed_mode = channel_ < 8 ? LEDC_HIGH_SPEED_MODE : LEDC_LOW_SPEED_MODE;
+  auto speed_mode = get_speed_mode(channel_);
   auto timer_num = static_cast<ledc_timer_t>((channel_ % 8) / 2);
   auto chan_num = static_cast<ledc_channel_t>(channel_ % 8);
 
@@ -114,7 +126,7 @@ void LEDCOutput::update_frequency(float frequency) {
     ESP_LOGW(TAG, "LEDC output hasn't been initialized yet!");
     return;
   }
-  auto speed_mode = channel_ < 8 ? LEDC_HIGH_SPEED_MODE : LEDC_LOW_SPEED_MODE;
+  auto speed_mode = get_speed_mode(channel_);
   auto timer_num = static_cast<ledc_timer_t>((channel_ % 8) / 2);
 
   ledc_timer_config_t timer_conf{};

